anomalib/utils/callbacks/selective_feature_model.py

```diff
@@ -0,0 +1,148 @@
+
+        if self.max_features is None:
+            warn("`self.max_features is None")
+            return
+
+        logger.info("Fitting Selective Feature model")
+
+        self.feature_model.fit(self.max_features, self.class_labels)
+
+        # reset arrays
+        self.max_features = None
+        self.class_labels = []
+
+    def on_test_epoch_start(self, _trainer: Trainer, _pl_module: LightningModule) -> None:
+        """Reset max_features and class_labels before testing starts."""
+        self.max_features = None
+        self.class_labels = []
+
+    def on_test_batch_end(
+        self,
+        _trainer: Trainer,
+        _pl_module: LightningModule,
+        outputs: Dict[str, Tensor],
+        _batch: Any,
+        _batch_idx: int,
+        _dataloader_idx: int,
+    ) -> None:
+        """Stores max activation values for all images.
+
+        Args:
+            outputs (Dict[str, Tensor]): Outputs containing `max_activation_val`.
+        """
+        if "max_activation_val" not in outputs.keys() and "class" not in outputs.keys():
+            warn("Need both max_activation_val and class keys in outputs. Skipping SFM test step")
+        else:
+            max_val_features = torch.vstack(list(outputs["max_activation_val"]))
+            class_labels = np.hstack(list(outputs["class"]))
+
+            if self.max_features is None:
+                self.max_features = max_val_features
+                self.class_labels = class_labels
+            else:
+                self.max_features = torch.vstack([self.max_features, max_val_features])
+                self.class_labels = np.hstack([self.class_labels, class_labels])
+
+    def on_test_epoch_end(self, _trainer: Trainer, pl_module: LightningModule) -> None:
+        """Compute sub-class testing accuracy."""
+        if self.max_features is None:
+            warn("`self.max_features is None")
+            return
+
+        class_names = np.unique(self.class_labels)
+
+        results: Dict[str, list] = {}
+        for class_name in class_names:
+            results[class_name] = []
+        # sorted values and idx for entire feature set
+        max_val, max_idx = torch.sort(self.max_features, descending=True)
+        reduced_range = int(max_val.shape[1] * 0.10)
+        # indexes of top 10% FEATURES HAVING MAX VALUE
+        top_max_idx = max_idx[:, 0:reduced_range]
+        correct_class = []
+        for idx, feature in enumerate(top_max_idx):
+            scores = []
+            for class_name in class_names:
+                stats = getattr(self.feature_model, class_name).cpu()
+                score = stats[1][np.isin(stats[0], feature)].sum()
+                scores.append(score)
+            scores[np.where(class_names == "good")[0][0]] = 0
+            if "combined" in class_names:
+                scores[np.where(class_names == "combined")[0][0]] = 0
+            if "thread" in class_names:
+                scores[np.where(class_names == "thread")[0][0]] = 0
```

Comment:
Do you know why we have this sub-category checks only for `good, thread and combined`?